Build failure: libmd

### Steps To Reproduce

Steps to reproduce the behavior:
1. build libmd

### Build log

```
--- sha512c.o ---
-error=int-in-bool-context -Wno-error=memset-elt-size -Wno-error=noexcept-type -Wno-error=nonnull -Wno-error=pointer-compare -Wno-error=stringop-overflow -Wno-error=aggressive-loop-optimizations -Wno-error=cast-function-type -Wno-error=catch-value -Wno-error=multistatement-macros -Wno-error=restrict -Wno-error=sizeof-pointer-memaccess -Wno-error=stringop-truncation -Wno-return-type -Wno-system-headers -Wno-system-headers -Wno-address-of-packed-member  -Wno-unused-result    -I/legacy/usr/include -Werror=implicit-function-declaration -Werror=implicit-int  -Werror=return-type -Wundef -DHAVE_NBTOOL_CONFIG_H=1 -I/build/libmd-filtered-src/tools/build/cross-build/include/linux -D_GNU_SOURCE=1 -I/build/libmd-filtered-src/tools/build/cross-build/include/common -c /build/libmd-filtered-src/sys/crypto/sha2/sha512c.c -o sha512c.o
/build/libmd-filtered-src/sys/crypto/sha2/sha512c.c:45:10: fatal error: machine/ifunc.h: No such file or directory
   45 | #include <machine/ifunc.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
*** [sha512c.o] Error code 1
```

### Additional context

Found when trying to use `pkgsCross.x86_64-freebsd` on aarch64-linux.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@primeos

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"aarch64-linux"`
 - host os: `Linux 6.8.12, NixOS, 24.05 (Uakari), 24.05.20240719.0c53b6b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - nixpkgs: `/home/ross/.cache/nixpkgs-review/pr-335703/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/os-specific/bsd/freebsd/pkgs/libelf.nix
+++ b/pkgs/os-specific/bsd/freebsd/pkgs/libelf.nix
@@ -23,6 +23,4 @@ mkDerivation {
     install
     m4
   ];
-
-  BOOTSTRAPPING = !stdenv.hostPlatform.isFreeBSD;
 }

--- a/pkgs/os-specific/bsd/freebsd/pkgs/localedef.nix
+++ b/pkgs/os-specific/bsd/freebsd/pkgs/localedef.nix
@@ -35,5 +35,4 @@ mkDerivation (
 
     MK_TESTS = "no";
   }
-  // lib.optionalAttrs (!stdenv.hostPlatform.isFreeBSD) { BOOTSTRAPPING = 1; }
 )

--- a/pkgs/os-specific/bsd/freebsd/pkgs/localedef.nix
+++ b/pkgs/os-specific/bsd/freebsd/pkgs/localedef.nix
@@ -9,30 +9,28 @@
   makeMinimal,
   install,
 }:
-mkDerivation (
-  {
-    path = "usr.bin/localedef";
+mkDerivation ({
+  path = "usr.bin/localedef";
 
-    extraPaths = [
-      "lib/libc/locale"
-      "lib/libc/stdtime"
-    ] ++ lib.optionals (!stdenv.hostPlatform.isFreeBSD) [ "." ];
+  extraPaths = [
+    "lib/libc/locale"
+    "lib/libc/stdtime"
+  ] ++ lib.optionals (!stdenv.hostPlatform.isFreeBSD) [ "." ];
 
-    nativeBuildInputs = [
-      bsdSetupHook
-      byacc
-      freebsdSetupHook
-      makeMinimal
-      install
-    ];
+  nativeBuildInputs = [
+    bsdSetupHook
+    byacc
+    freebsdSetupHook
+    makeMinimal
+    install
+  ];
 
-    buildInputs = [ ];
+  buildInputs = [ ];
 
-    preBuild = lib.optionalString (!stdenv.hostPlatform.isFreeBSD) ''
-      export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -I${compat}/include -D__unused= -D__pure= -Wno-strict-aliasing"
-      export NIX_LDFLAGS="$NIX_LDFLAGS -L${compat}/lib"
-    '';
+  preBuild = lib.optionalString (!stdenv.hostPlatform.isFreeBSD) ''
+    export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -I${compat}/include -D__unused= -D__pure= -Wno-strict-aliasing"
+    export NIX_LDFLAGS="$NIX_LDFLAGS -L${compat}/lib"
+  '';
 
-    MK_TESTS = "no";
-  }
-)
+  MK_TESTS = "no";
+})

--- a/pkgs/os-specific/bsd/freebsd/pkgs/mkDerivation.nix
+++ b/pkgs/os-specific/bsd/freebsd/pkgs/mkDerivation.nix
@@ -92,6 +92,7 @@ lib.makeOverridable (
       # Since STRIP in `makeFlags` has to be a flag, not the binary itself
       STRIPBIN = "${stdenv'.cc.bintools.targetPrefix}strip";
     }
+    // lib.optionalAttrs (!stdenv.hostPlatform.isFreeBSD) { BOOTSTRAPPING = true; }
     // lib.optionalAttrs stdenv'.hostPlatform.isDarwin { MKRELRO = "no"; }
     // lib.optionalAttrs (stdenv'.cc.isClang or false) {
       HAVE_LLVM = lib.versions.major (lib.getVersion stdenv'.cc.cc);

--- a/pkgs/os-specific/bsd/freebsd/pkgs/mkDerivation.nix
+++ b/pkgs/os-specific/bsd/freebsd/pkgs/mkDerivation.nix
@@ -56,8 +56,6 @@ lib.makeOverridable (
       ] ++ attrs.extraNativeBuildInputs or [ ];
       buildInputs = compatIfNeeded;
 
-      HOST_SH = stdenv'.shell;
-
       makeFlags =
         [
           "STRIP=-s" # flag to install, not command
@@ -65,14 +63,36 @@ lib.makeOverridable (
         ++ lib.optional (!stdenv'.hostPlatform.isFreeBSD) "MK_WERROR=no"
         ++ lib.optional stdenv.hostPlatform.isStatic "SHLIB_NAME=";
 
-      # amd64 not x86_64 for this on unlike NetBSD
-      MACHINE_ARCH = freebsd-lib.mkBsdArch stdenv';
+      env =
+        {
+          HOST_SH = stdenv'.shell;
 
-      MACHINE = freebsd-lib.mkBsdMachine stdenv';
+          # amd64 not x86_64 for this on unlike NetBSD
+          MACHINE_ARCH = freebsd-lib.mkBsdArch stdenv';
 
-      MACHINE_CPUARCH = freebsd-lib.mkBsdCpuArch stdenv';
+          MACHINE = freebsd-lib.mkBsdMachine stdenv';
 
-      COMPONENT_PATH = attrs.path or null;
+          MACHINE_CPUARCH = freebsd-lib.mkBsdCpuArch stdenv';
+
+          COMPONENT_PATH = attrs.path or null;
+        }
+        // lib.optionalAttrs stdenv'.hasCC {
+          # TODO should CC wrapper set this?
+          CPP = "${stdenv'.cc.targetPrefix}cpp";
+
+          # Since STRIP in `makeFlags` has to be a flag, not the binary itself
+          STRIPBIN = "${stdenv'.cc.bintools.targetPrefix}strip";
+        }
+        // lib.optionalAttrs (!stdenv.hostPlatform.isFreeBSD) { BOOTSTRAPPING = true; }
+        // lib.optionalAttrs stdenv'.hostPlatform.isDarwin { MKRELRO = "no"; }
+        // lib.optionalAttrs (stdenv'.cc.isClang or false) {
+          HAVE_LLVM = lib.versions.major (lib.getVersion stdenv'.cc.cc);
+        }
+        // lib.optionalAttrs (stdenv'.cc.isGNU or false) {
+          HAVE_GCC = lib.versions.major (lib.getVersion stdenv'.cc.cc);
+        }
+        // lib.optionalAttrs (stdenv'.hostPlatform.isx86_32) { USE_SSP = "no"; }
+        // (attrs.env or { });
 
       strictDeps = true;
 
@@ -85,27 +105,11 @@ lib.makeOverridable (
         license = lib.licenses.bsd2;
       } // attrs.meta or { };
     }
-    // lib.optionalAttrs stdenv'.hasCC {
-      # TODO should CC wrapper set this?
-      CPP = "${stdenv'.cc.targetPrefix}cpp";
-
-      # Since STRIP in `makeFlags` has to be a flag, not the binary itself
-      STRIPBIN = "${stdenv'.cc.bintools.targetPrefix}strip";
-    }
-    // lib.optionalAttrs (!stdenv.hostPlatform.isFreeBSD) { BOOTSTRAPPING = true; }
-    // lib.optionalAttrs stdenv'.hostPlatform.isDarwin { MKRELRO = "no"; }
-    // lib.optionalAttrs (stdenv'.cc.isClang or false) {
-      HAVE_LLVM = lib.versions.major (lib.getVersion stdenv'.cc.cc);
-    }
-    // lib.optionalAttrs (stdenv'.cc.isGNU or false) {
-      HAVE_GCC = lib.versions.major (lib.getVersion stdenv'.cc.cc);
-    }
-    // lib.optionalAttrs (stdenv'.hostPlatform.isx86_32) { USE_SSP = "no"; }
     // lib.optionalAttrs (attrs.headersOnly or false) {
       installPhase = "includesPhase";
       dontBuild = true;
     }
-    // attrs
+    // (builtins.removeAttrs attrs [ "env" ])
     // lib.optionalAttrs (stdenv'.hasCC && stdenv'.cc.isClang or false && attrs.clangFixup or true) {
       preBuild =
         ''

--- a/pkgs/os-specific/bsd/freebsd/pkgs/mkcsmapper.nix
+++ b/pkgs/os-specific/bsd/freebsd/pkgs/mkcsmapper.nix
@@ -13,8 +13,6 @@ mkDerivation {
     "lib/libiconv_modules/mapper_std"
   ];
 
-  BOOTSTRAPPING = !stdenv.hostPlatform.isFreeBSD;
-
   extraNativeBuildInputs = [
     byacc
     flex

--- a/pkgs/os-specific/bsd/freebsd/pkgs/mkesdb.nix
+++ b/pkgs/os-specific/bsd/freebsd/pkgs/mkesdb.nix
@@ -1,5 +1,4 @@
 {
-  stdenv,
   mkDerivation,
   byacc,
   flex,
@@ -9,8 +8,6 @@ mkDerivation {
   path = "usr.bin/mkesdb";
 
   extraPaths = [ "lib/libc/iconv" ];
-
-  BOOTSTRAPPING = !stdenv.hostPlatform.isFreeBSD;
 
   extraNativeBuildInputs = [
     byacc

